pgmax/factors/enumeration.py

```diff
@@ -176,60 +152,115 @@ def concatenate_wirings(wirings: Sequence[EnumerationWiring]) -> EnumerationWiri
-            of its n variable states are m, m + 1, ..., m + n - 1
-        num_factors: Number of Factors in the FactorGroup.
+        Raises:
+            ValueError: if factor_edges_num_states is not of shape (num_factors * num_variables, )
 
-    Returns:
-        The EnumerationWiring
-    """
-    var_states_for_edges = []
-    for variables_for_factor in variables_for_factors:
-        for variable in variables_for_factor:
-            num_states = variable.num_states
-            this_var_states_for_edges = np.arange(
-                vars_to_starts[variable], vars_to_starts[variable] + num_states
+        Returns:
+            The EnumerationWiring
+        """
+        var_states = np.array(
+            [vars_to_starts[variable] for variable in variables_for_factors]
+        )
+        num_states = np.array(
+            [variable.num_states for variable in variables_for_factors]
+        )
+        num_states_cumsum = np.insert(np.cumsum(num_states), 0, 0)
+        var_states_for_edges = np.empty(shape=(num_states_cumsum[-1],), dtype=int)
+        _compile_var_states_numba(var_states_for_edges, num_states_cumsum, var_states)
+
+        num_configs, num_variables = factor_configs.shape
+        if not factor_edges_num_states.shape == (num_factors * num_variables,):
+            raise ValueError(
+                f"Expected factor_edges_num_states shape is {(num_factors * num_variables,)}. Got {factor_edges_num_states.shape}."
             )
-            var_states_for_edges.append(this_var_states_for_edges)
+        factor_configs_edge_states = np.empty(
+            (num_factors * num_configs * num_variables, 2), dtype=int
+        )
+        factor_edges_starts = np.insert(np.cumsum(factor_edges_num_states), 0, 0)
+        _compile_enumeration_wiring_numba(
+            factor_configs_edge_states, factor_configs, factor_edges_starts, num_factors
+        )
 
-    # Note: edges_starts corresponds to the factor_to_msgs_start for the LogicalFactors
-    edges_starts = np.insert(factor_edges_num_states.cumsum(), 0, 0)[:-1].reshape(
-        -1, factor_configs.shape[1]
-    )
+        return EnumerationWiring(
+            edges_num_states=factor_edges_num_states,
+            var_states_for_edges=var_states_for_edges,
+            factor_configs_edge_states=factor_configs_edge_states,
+        )
 
-    factor_configs_edge_states = np.stack(
-        [
-            np.repeat(
-                np.arange(factor_configs.shape[0] * num_factors),
-                factor_configs.shape[1],
-            ),
-            (factor_configs[None] + edges_starts[:, None, :]).flatten(),
-        ],
-        axis=1,
-    )
-    return EnumerationWiring(
-        edges_num_states=factor_edges_num_states,
-        var_states_for_edges=np.concatenate(var_states_for_edges),
-        factor_configs_edge_states=factor_configs_edge_states,
-    )
+
+@nb.jit(parallel=False, cache=True, fastmath=True, nopython=True)
+def _compile_var_states_numba(
```

Comment:
Is there a reason you make the caller allocate these arrays?  In general t's cleaner and less likely to result in error to allocate return arrays inside numba rather that mutating a passed in array - just do it outside of the for loop.  You can get a very small optimization by re-using arrays between calls (so highly performance sensitive code it can be useful), but you're not doing that here.